File: problem_30/problem_30.py
# first of all, we have to find the upper bound of the number
# assume that the number is 9...9, then the upper bound is 9^5 * n = 10^n

upper = 1
while 9**5 * upper > 10**upper:
    upper += 1

print(upper) # 6, so the upper bound is 10^6 = 1000000

# then we can start to find the number

def digit_fifth_powers(n):
    lst = [] # for listing the numbers
    n = 10**n

    for i in range(2, n):
        sum = 0
        temp = i # for storing the number
        while temp > 0:
            sum += (temp % 10)**5
            temp //= 10
        if sum == i:
            lst.append(i)
    return lst

number_list = digit_fifth_powers(upper) # find the numbers

print(number_list) # [4150, 4151, 54748, 92727, 93084, 194979]

sum_of_numbers = 0

for i in number_list:
    sum_of_numbers += i

print(sum_of_numbers) # 443839

File: problem_30/problem_30.c
#include <stdio.h>

// in python file, we found that the upper bound is 1000000

int main(){

    int sum_of_numbers = 0;

    for (int i = 2; i < 1000000; i++){

        int sum = 0;
        int number = i; // we need to keep the original number

        while (number > 0){
            int digit = number % 10;
            sum += digit * digit * digit * digit * digit;
            number /= 10;
        }

        if (sum == i){
            sum_of_numbers += i;
        }
    }

    printf("%d\n", sum_of_numbers);
    return 0;

}

//answer: 443839